python/ray/data/datasource/file_meta_provider.py

```diff
@@ -46,3 +53,271 @@ def __call__(
+        *,
+        rows_per_file: Optional[int],
+        file_sizes: List[Optional[int]],
+    ) -> BlockMetadata:
+        if rows_per_file is None:
+            num_rows = None
+        else:
+            num_rows = len(paths) * rows_per_file
+        return BlockMetadata(
+            num_rows=num_rows,
+            size_bytes=None if None in file_sizes else sum(file_sizes),
+            schema=schema,
+            input_files=paths,
+            exec_stats=None,
+        )  # Exec stats filled in later.
+
+    def expand_paths(
+        self,
+        paths: List[str],
+        filesystem: "pyarrow.fs.FileSystem",
+    ) -> Tuple[List[str], List[Optional[int]]]:
+        from pyarrow.fs import FileType
+        from ray.data.datasource.file_based_datasource import _expand_directory
+
+        if len(paths) > 1:
+            logger.warning(
+                f"Expanding {len(paths)} path(s). This may be a HIGH LATENCY "
+                f"operation on some cloud storage services. If the specified paths "
+                f"all point to files and never directories, try rerunning this read "
+                f"with `meta_provider=FastFileMetadataProvider()`."
+            )
+        expanded_paths = []
+        file_infos = []
+        for path in paths:
+            file_info = filesystem.get_file_info(path)
+            if file_info.type == FileType.Directory:
+                paths, file_infos_ = _expand_directory(path, filesystem)
+                expanded_paths.extend(paths)
+                file_infos.extend(file_infos_)
+            elif file_info.type == FileType.File:
+                expanded_paths.append(path)
+                file_infos.append(file_info)
+            else:
+                raise FileNotFoundError(path)
+        file_sizes = [file_info.size for file_info in file_infos]
+        return expanded_paths, file_sizes
+
+
+class FastFileMetadataProvider(DefaultFileMetadataProvider):
+    """Fast Metadata provider for FileBasedDatasource implementations.
+
+    Offers improved performance vs. DefaultFileMetadataProvider by skipping directory
+    path expansion and file size collection. While this performance improvement may be
+    negligible for local filesystems, it can be substantial for cloud storage service
+    providers.
+
+    This should only be used when all input paths are known to be files.
+    """
+
+    def expand_paths(
+        self,
+        paths: List[str],
+        filesystem: "pyarrow.fs.FileSystem",
+    ) -> Tuple[List[str], List[Optional[int]]]:
+        logger.warning(
+            f"Skipping expansion of {len(paths)} path(s). If your paths contain "
+            f"directories or if file size collection is required, try rerunning this "
+            f"read with `meta_provider=DefaultFileMetadataProvider()`."
+        )
+        import numpy as np
+
+        return paths, np.empty(len(paths), dtype=object)
+
+
+@DeveloperAPI
+class ParquetMetadataProvider(FileMetadataProvider):
```

Comment:
I have similar concern of whether this layer is needed, i.e. can DefaultParquetMetadataProvider directly subclass FileMetadataProvider?

Reply:
I have a slight preference to preserve the current layout since the intent is for anyone providing a custom `ParquetDatasource` metadata provider to implement the interface signature established in `ParquetMetadataProvider` and, in particular:
1. Only implement metadata prefetching if their use-case requires it. 
2. Carefully consider the best way to fetch block metadata for their use-case rather than just keep a default implementation.

For example, my `CachedFileMetadataProvider` at https://github.com/pdames/deltacat/blob/edec6159c5acda3ede15653b4e92aaa45a43206f/deltacat/io/aws/redshift/redshift_datasource.py#L67-L70 inherits from `ParquetMetadataProvider`, does not require metadata prefetching, and uses a different implementation for getting block metadata from a prebuilt cache. I also expect this to be the general case for most upcoming data warehouse and data catalog integrations with Ray Datasets beyond just the Amazon Redshift integration that uses it here.

So, in summary, my slight preference is to keep `DefaultParquetMetadataProvider` set aside as an internal implementation detail exposed to Ray Data maintainers (and to continue to keep the `@DeveloperAPI` label excluded from this class), and for `ParquetMetadataProvider` to be exposed to end-users creating their own `ParquetDatasource` metadata providers.

Reply:
SG, thanks.